ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
+			rc = memcached_server_add(mc, StringValueCStr(rb_hostname), NUM2INT(rb_port));
+			rb_memcached_error_check(rc);
+		}
+		else if (SYM2ID(rb_backend) == id_socket) {
+			VALUE rb_fd = rb_ary_entry(rb_server, 1);
+			Check_Type(rb_fd, T_STRING);
+			rc = memcached_server_add_unix_socket(mc, StringValueCStr(rb_fd));
+			rb_memcached_error_check(rc);
+		}
+	}
+
+	// Verify keys by default
+	rc = memcached_behavior_set(mc, MEMCACHED_BEHAVIOR_VERIFY_KEY, true);
+	rb_memcached_error_check(rc);
+
+	return rb_mc;
 }
 
 static memcached_return_t
 iterate_server_function(const memcached_st *ptr, const memcached_instance_st * server, void *context)
 {
-  VALUE server_list = (VALUE) context;
-
-  VALUE rb_server = rb_obj_alloc(rb_cServer);
-
-  rb_ivar_set(rb_server, id_ivar_hostname, rb_str_new2(memcached_server_name(server)));
-  rb_ivar_set(rb_server, id_ivar_port, UINT2NUM(memcached_server_port(server)));
-  // TODO weight
+	VALUE rb_server_list = (VALUE)context;
+	VALUE rb_server = rb_funcall(rb_cServer, rb_intern("new"), 2, 
+			rb_str_new2(memcached_server_name(server)),
+			UINT2NUM(memcached_server_port(server)));
 
-  rb_ary_push(server_list, rb_server);
-  return MEMCACHED_SUCCESS;
+	rb_ary_push(rb_server_list, rb_server);
+	return MEMCACHED_SUCCESS;
 }
 
 static VALUE
 rb_connection_servers(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
+	memcached_server_fn callbacks[1] = { &iterate_server_function };
+	memcached_st *mc;
 
-  VALUE server_list = rb_ary_new();
+	VALUE rb_result;
 
-  memcached_server_fn callbacks[1];
-  callbacks[0] = iterate_server_function;
+	UnwrapMemcached(self, mc);
+	rb_result = rb_ary_new();
 
-  memcached_server_cursor(ctx->memc, callbacks, (void *)server_list, 1);
-  return server_list;
+	memcached_server_cursor(mc, callbacks, (void *)rb_result, 1);
+	return rb_result;
 }
 
 static VALUE
 rb_connection_flush(VALUE self)
 {
-  memcached_ctx *ctx = get_ctx(self);
+	memcached_st *mc;
+	memcached_return_t rc;
 
-  memcached_return_t rc = memcached_flush(ctx->memc, 0);
-  handle_memcached_return(rc);
+	UnwrapMemcached(self, mc);
+	rc = memcached_flush(mc, 0);
+	rb_memcached_error_check(rc);
 
-  return (rc == MEMCACHED_SUCCESS);
+	return (rc == MEMCACHED_SUCCESS) ? Qtrue : Qfalse;
 }
 
 static VALUE
-rb_connection_set(VALUE self, VALUE key, VALUE value, VALUE ttl, VALUE flags)
+rb_connection_set(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(key);
-  char *mvalue = StringValuePtr(value);
-
-  memcached_return_t rc = memcached_set(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(ttl), NUM2INT(flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_set(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
```

Comment:
Instead, we're going to use `RSTRING_PTR` and `RSTRING_LEN` throughout the wrapper, to ensure we're passing the whole string. No `strlen` to be seen here! Remember we can be dealing with strings with embedded NULL bytes (which would cause `strlen` to report a wrong length).

Also note that since the `RSTRING_` helpers are macros, we need to be very careful with typechecking or we will crash.